air-quality-ui/src/components/single-city/station-map/StationMap.tsx

```diff
@@ -1,44 +1,117 @@
+  Marker,
+  MarkerOptions,
+  Popup,
+} from 'maplibre-gl'
 import { useEffect, useRef } from 'react'
+import { createRoot } from 'react-dom/client'
 
 import { createMapConfig } from './map-service'
 import classes from './StationMap.module.css'
 import 'maplibre-gl/dist/maplibre-gl.css'
+import { StationPopup } from './StationPopup'
 import { Coordinates } from '../../../services/types'
 
+export type Station = { name: string; latitude: number; longitude: number }
+
 interface AverageComparisonChartProps {
   mapCenter: Coordinates
-  locations: Map<string, { longitude: number; latitude: number }>
+  stations: Record<string, Station>
+  visibleLocations: string[]
+  stationColors: Record<string, string>
+  removeSite: (siteName: string) => void
+  addSite: (siteName: string) => void
+}
+
+const createPopup = (
+  stationName: string,
+  removeSite: (id: string) => void,
+  addSite: (id: string) => void,
+  remove: boolean = true,
+) => {
+  const div = document.createElement('div')
+  const root = createRoot(div)
+  root.render(
+    <StationPopup
+      stationName={stationName}
+      removeSite={(id) => removeSite(id)}
+      addSite={(id) => addSite(id)}
+      remove={remove}
+    />,
+  )
+  return new Popup().setDOMContent(div)
 }
 
 export const StationMap = (props: AverageComparisonChartProps) => {
   const mapContainer = useRef<HTMLDivElement>(null)
-  const map = useRef<maplibregl.Map | null>(null)
-  const mapCenterLongitude = props.mapCenter.longitude
-  const mapCenterLatitude = props.mapCenter.latitude
-  const zoom = 9
+  const map = useRef<MapMap | null>(null)
+  const markers = useRef(new Map<string, Marker>())
 
   useEffect(() => {
-    const mapconfig = new maplibregl.Map(
+    const zoom = 9
+    const mapconfig = new MapMap(
       createMapConfig(
         mapContainer.current!,
-        mapCenterLatitude,
-        mapCenterLongitude,
+        props.mapCenter.latitude,
+        props.mapCenter.longitude,
         zoom,
       ),
     )
 
     mapconfig.addControl(new FullscreenControl())
 
-    props.locations.forEach((value, name) => {
-      new Marker()
-        .setLngLat([value.longitude, value.latitude])
-        .setPopup(new Popup().setText(name))
-        .addTo(mapconfig)
+    map.current = mapconfig
+  }, [props.mapCenter.latitude, props.mapCenter.longitude])
+
+  const { stations, removeSite, addSite } = props
+
+  useEffect(() => {
+    Object.values(stations).forEach((station) => {
+      const marker = new Marker({ color: props.stationColors[station.name] })
+      marker.setLngLat([station.longitude, station.latitude])
+      marker.setPopup(createPopup(station.name, removeSite, addSite))
+      marker.addTo(map.current!)
+
+      markers.current?.set(station.name, marker)
     })
+  }, [stations, props.stationColors, removeSite, addSite])
 
-    map.current = mapconfig
-  }, [mapCenterLongitude, mapCenterLatitude, zoom, props.locations])
+  const firstLoad = useRef<boolean>(true)
+
+  useEffect(() => {
```

Comment:
I might be missing something but can we combine this to a single effect if we add props.visibleLocations to the above?

Reply:
don't really want to re-run the other useEffect code when the visibleLocations prop updates. In fact it should only ever run once so I've merged it with the useEffect that creates the map